### DISCUSSION TITLE
Xonsh as shell interaction library

### DISCUSSION THREAD

**Discussion by ShalokShalom** [xonsh/xonsh#5090] [Q&A] [answered]
Hi there :)

Your webpage, and the Github README suggests to me that I can use Xonsh in order to interact between Python and the OS. 

The webpage directs to the tutorial, which goes on and says nothing about that use case. 

I have not found documentation to do so at all, and I think, if this is possible, that this deserves to be in the comparison sheet with other shell languages. 

To the details:
I am in the midst of writing a GUI layer on top of machinectl, mkiso, and other systemd-nspawn related commands.

For that, I need to call the binaries and detect their return value. I see no tutorial, that would show that. I will then like to further use those values in Python. 

Thanks

**Comment by bestlem** [comment]
I think you have misunderstood xonsh. It is a shell - to replace PowerShell, cmd, bash, zsh etc.
It allows you to call programs from the command line or xonsh scripts easily.

I think what you need to look at is in the standard python libraries. The linrary is the [os module](https://docs.python.org/3/library/os.html) especially [Process Management](https://docs.python.org/3/library/os.html#process-management). xonsh does not provide much for python programs - internally it uses these fuctions.

**Reply by ShalokShalom** [reply]
I understood, that this is, what Xonsh is mostly. And still the website suggests that my use case is also covered: 

![Screenshot_20230317_120501](https://user-images.githubusercontent.com/6344099/225887736-d146e8bf-2fcf-48fa-9c14-6dc01d453fe6.png)

Is there some way to inherit a Xonsh script in Python?

**Reply by bestlem** [reply]
Just make your app an xonsh app rather than a python one

ie the script starts with `#!/usr/bin/env xonsh` rather than python.
But that makes other tools e.g. IDEs harder to use. I think the example above is only useful for simple scripting. os or the subprocess module I think would be better.

**Reply by anki-code** [reply]
The answer moved to thread - https://github.com/xonsh/xonsh/discussions/5090#discussioncomment-5344738

**Reply by ShalokShalom** [reply]
So I basically have to use Xonsh, and not Python? That's fine to me. I like, at some point, direct all that into Python, nonetheless. Like to create the QML interface, as an example. How do I do that?

**Comment by anki-code** [comment, marked as answer]
@ShalokShalom please start from [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet) - it has good description of key aspects and most of use cases in the nutshell with copy-pastable examples.

In short if you have for example `echo` (or `machinectl`) and you want to write a wrapper for example as [REST API on flask](https://pythonbasics.org/flask-rest-api/). It works like a charm:
```xsh
xonsh
xpip install flask
cd /tmp

echo @("""

import json
from flask import Flask
app = Flask(__name__)
@app.route('/echo')
def index():
    result = $(echo -n hello from echo)  # run subprocess command
    return json.dumps({'result': result})
app.run()

""") > myapi.xsh
chmod +x myapi.xsh

./myapi.xsh
# Running on http://127.0.0.1:5000

curl http://127.0.0.1:5000/echo
# {"result": "hello from echo"}
```
Also xonsh is Python library. And you can call xonsh functions from pure Python i.e.:
```xsh
# https://xon.sh/api/_autosummary/cmd/xonsh.built_ins.html#xonsh.built_ins.subproc_captured_stdout
from xonsh.built_ins import subproc_captured_stdout as run
result = run(['echo', 'hello'])
print(repr(result))
# 'hello\n'
```

**Reply by ShalokShalom** [reply]
I think this deserves better documentation 

The usage of Xonsh as Python library, seem to be undocumented

**Reply by anki-code** [reply]
Xonsh lib API documented here - https://xon.sh/api/index.html

**Reply by PrestonMoak9** [reply]
Other folks in this thread have chimed in with some helpful tips and resources, so hopefully you're feeling a bit more confident about how to proceed. And don't forget that the Xonsh community is always here to help if you run into any more roadblocks along the way.

**Reply by ShalokShalom** [reply]
Hi. 

Thanks for your input. 

I suggest to rethink the documentation (features who are not documented, or where documentation is hard to find, could as well not exit) in relationship with the advertising. These two things dont match, and I still think this project could do a much better job, both in terms of ergonomics of this feature, and as well by its documentation. 

This could fill a hole in the ecosystem. 🌏 

Greatly appreciated 🥳

**Reply by anki-code** [reply]
PRs to improve documentation are welcome!
